docs/proposals/20191017-control-plane.md

```diff
@@ -0,0 +1,495 @@
+}
+
+```
+
+With the following validations:
+
+- If the referenced InitTemplate does not define external etcd (webhook):
+  - ControlPlane.Spec.Replicas in [1, 3, 5, 7]
+- ControlPlane.Spec.Version != “” (openapi)
+
+And the following defaulting:
+
+- ControlPlaneSpec.Replicas: 1 (openapi or webhook)
+
+#### Modifications required to existing API Types
+
+- Add Cluster.Spec.ControlPlaneRef defined as:
+
+```go
+    // ControlPlaneRef is an optional reference to a provider-specific resource that holds
+    // the details for provisioning the Control Plane for a Cluster
+    // +optional
+    ControlPlaneRef *corev1.ObjectReference `json:"controlPlaneRef,omitempty"`
+```
+
+- Add Cluster.Status.ControlPlaneReady defined as:
+
+```go
+    // ControlPlaneReady defines if the control plane is ready
+    // +optional
+    ControlPlaneReady bool `json:"controlPlaneReady"`
+```
+
+#### Behavioral Changes from v1alpha2
+
+- If Cluster.Spec.ControlPlaneRef is set:
+  - [Status.ControlPlaneInitialized](https://github.com/kubernetes-sigs/cluster-api/issues/1243) is set based on the value of Status.Initialized for the referenced resource
+  - Status.ControlPlaneReady is set based on the value of Status.Ready for the referenced resource, this field is intended to eventually replace Status.ControlPlaneInitialized as a field that will be kept up to date instead of set only once.
+- Current behavior will be preserved if Cluster.Spec.ControlPlaneRef is not set
+
+#### Behaviors
+
+##### Create
+
+- A created ControlPlane should bootstrap a control plane with a given number of replicas
+- If an error occurs, errorStatus and errorReason will be defined on the ControlPlane object
+- Cannot create an even number of stacked control planes or anything other than 1, 3, 5, or 7
+- Can create an arbitrary number of control planes if etcd is external to the control plane
+- Creating a ControlPlane with > 1 replicas is equivalent to creating a ControlPlane with 1 replica followed by scaling the ControlPlane to 3 replicas
+- The kubeadm bootstrapping configuration provided via ControlPlane.Spec.KubeadmConfigSpec, should specify the initConfiguration, clusterConfiguration, and joinConfiguration stanzas, and the ControlPlane controller will be responsible for splitting the config and passing it to the underlying Machines created as appropriate
+- The underlying query used to find existing Control Plane Machines is based on the following hardcoded label selector:
+
+```yaml
+  selector:
+    matchLabels:
+      cluster.x-k8s.io/cluster-name: my-cluster
+      cluster.x-k8s.io/controlplane: "true"
+```
+
+- Generate CA certificates if this is the first control plane machine in the cluster
+- Generate the kubeconfig secret if this is the first control plane machine in the cluster
+
+Given the following `cluster.yaml` file:
+
+```yaml
+---
+kind: Cluster
+apiVersion: cluster.x-k8s.io/v1alpha3
+metadata:
+  name: my-cluster
+  namespace: default
+Spec:
+    clusterNetwork:
```

Comment:
Is clusterNetwork mainly to generate controlPlane kubeadmConfig? If so, is it a more natural fit to move under ControlPlaneSpec?

Reply:
The cluster network (i.e. the pod network) is used by the data plane, although in some cases, control plane component (kube-controller-manager) needs to know the value, too.

For reference, see https://kubernetes.io/docs/setup/production-environment/tools/kubeadm/create-cluster-kubeadm/#pod-network.

Reply:
yeah, but afaik, the definition here is only used to generate ClusterConfiguration for kubeadmConfig, which only applies to ControlPlane. Only exception is APIServerPort which is used by AWS provider to create elb.

Reply:
With the *default* implementations this is only currently used in those places. I think it's a valuable conversation to have if having ClusterNetwork defined as part of the Cluster Spec still has value to the project, I'm not sure I want to take that on as part of this proposal, though.

Reply:
sure. thanks.